### PR TITLE
Solution to : sem.05/practicum/Date

### PR DESCRIPTION
**Pull request by JustImprove**

### CODE REVIEW

File: Sem. 05/Practicum/Dates.h [hunk 1]
```diff
@@ -0,0 +1,39 @@
...
+    September, 
+    October, 
+    November, 
+    December
+};
+class Date 
+{
+    public: 
+        Date();
+        Date(int Day, Months Month, int Year);
```
Comment (Justsvetoslavov) [inline Sem. 05/Practicum/Dates.h:22]: int vs enum. Data type is int - change req.

File: Sem. 05/Practicum/Dates.h [hunk 2]
```diff
@@ -0,0 +1,39 @@
...
+        Date(int Day, Months Month, int Year); 
+        void addDays(int DaysCount); 
+        int getDays() const { return Day; };
+        int getMonth() const { return Month; };
+        int getYear() const { return Year; };
+        void removeDays(int DaysCount); 
+        int daysToXmas();
+        bool isLeapYear(int Year = 0) const;
+        bool isEarlierThan(const Date& date) const;
+        int daysToEndOfTheYear();
```
Comment (Justsvetoslavov) [inline Sem. 05/Practicum/Dates.h:31]: long long daysToNewYear() const;

File: Sem. 05/Practicum/Dates.h [hunk 3]
```diff
@@ -0,0 +1,39 @@
...
+    public: 
+        Date();
+        Date(int Day, Months Month, int Year); 
+        void addDays(int DaysCount); 
+        int getDays() const { return Day; };
+        int getMonth() const { return Month; };
+        int getYear() const { return Year; };
+        void removeDays(int DaysCount); 
+        int daysToXmas();
+        bool isLeapYear(int Year = 0) const;
```
Comment (Justsvetoslavov) [inline Sem. 05/Practicum/Dates.h:29]: remove parameter.

File: Sem. 05/Practicum/Dates.h [hunk 4]
```diff
@@ -0,0 +1,39 @@
...
+        bool isEarlierThan(const Date& date) const;
+        int daysToEndOfTheYear();
+        void printDate() const;
+        int daysUntilCurrentDate(const Date& date);
```
Comment (Justsvetoslavov) [inline Sem. 05/Practicum/Dates.h:33]: long long daysBetweenDates(const Date&) const;

File: Sem. 05/Practicum/Dates.cpp [hunk 5]
```diff
@@ -0,0 +1,167 @@
+#include "Dates.h"
+#include <iostream>
+
+#pragma region Dates::methods 
+int Date::getDaysInMonth(int Month, int Year) const
```
Comment (Justsvetoslavov) [inline Sem. 05/Practicum/Dates.cpp:5]: Refactor required.

File: Sem. 05/Practicum/Dates.cpp [hunk 6]
```diff
@@ -0,0 +1,167 @@
...
+    }
+}
+Date::Date() : Day(1), Month(1), Year(2000) {};
+Date::Date(int Day, Months Month, int Year) 
+{
+    this->Day = Day;
+    this->Month = Month;
+    this->Year = Year;
+}
+bool Date::isLeapYear(int Year) const
```
Comment (Justsvetoslavov) [inline Sem. 05/Practicum/Dates.cpp:50]: Refactor , too many lines of code / too much return true/false . Do a 1 liner

File: Sem. 05/Practicum/Dates.cpp [hunk 7]
```diff
@@ -0,0 +1,167 @@
...
+            return false;
+        }
+        return true;
+    }
+    else
+    {
+        return false;
+    }
+}
+void Date::addDays(int DaysCount)
```
Comment (Justsvetoslavov) [inline Sem. 05/Practicum/Dates.cpp:70]: Refactor - avoid while(true).

File: Sem. 05/Practicum/Dates.cpp [hunk 8]
```diff
@@ -0,0 +1,167 @@
+#include "Dates.h"
+#include <iostream>
+
+#pragma region Dates::methods
```
Comment (Justsvetoslavov) [inline Sem. 05/Practicum/Dates.cpp:4]: Visual studio specific. Not needed.

File: Sem. 05/Practicum/Dates.cpp [hunk 9]
```diff
@@ -0,0 +1,167 @@
...
+}
+void Date::printDate() const 
+{
+    (Day < 10) ? std::cout << '0' << Day << '.' : std::cout << Day << '.';
+    (Month < 10) ? std::cout << '0' << Month << '.' : std::cout << Month << '.';
+    std::cout << Year << std::endl;
+}
+bool Date::isEarlierThan(const Date& date) const 
+{
+    if(Year < date.getYear() || Month < date.getMonth() || Day < date.getDays())
```
Comment (Justsvetoslavov) [inline Sem. 05/Practicum/Dates.cpp:97]: Example: 2000 < 1990 OR 4 < 5 -> Result is: False or True , which results in TRUE - Wrong Logic.

**Review by Justsvetoslavov — APPROVED** [review]

File: Sem. 05/Practicum/Date.cpp [hunk 10]
```diff
@@ -0,0 +1,154 @@
...
+const short DEFAULT_YEAR = 2000;
+
+//Constructors
+Date::Date()
+{
+	days = month = DEFAULT;
+	year = DEFAULT_YEAR;
+}
+
+Date::Date(const int _days, const int _month, const int _year)
```
Comment (YavorAlexandrov) [inline Sem. 05/Practicum/Date.cpp:14]: _Suggestion:_ Avoid starting variable names with underscores as they are usually saved for system variables.

**Review by YavorAlexandrov — APPROVED** [review]

**Pull request merged**